File: docs/JAMIA/comparisons/RPython/TM/38969267_select_columns_via_list.ts
/**
 * RPython SO#38969267 — Selecting columns via a list of column names
 * Effect: Crash
 * Bug class: Implicit column selection
 *
 * In pandas, df[list] with a non-existent column name crashes at runtime.
 *
 * In tidy-ts, select() only accepts column names that exist on the row type.
 */
import { createDataFrame } from "@tidy-ts/dataframe";

const df = createDataFrame([
  { a: 1, b: 4, c: 7 },
  { a: 2, b: 5, c: 8 },
  { a: 3, b: 6, c: 9 },
]);

const columnList = ["a", "b", "missing_col"] as const;

// @ts-expect-error — "missing_col" is not a key of the row type
const wrong = df.select(...columnList);

const correct = df.select("a", "b");

correct.print();

File: docs/JAMIA/comparisons/RPython/TM/38969267_select_columns_via_list.py
# ID: SO#38969267
# Language: Python
# Bug class: Column ref
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: Selecting columns via list fails when column doesn't exist. Column reference error.
# Reproduction status: Fixed (pandas 3)
# Type system catch: Schema type tracks column presence; non-existent key rejected

import pandas as pd

df = pd.DataFrame({
    "a": [1, 2, 3],
    "b": [4, 5, 6],
    "c": [7, 8, 9],
})

# SO question used the variable name `list` for the column-name list
list = ["a", "b", "missing_col"]

try:
    df_new = df[[list]]
except TypeError as e:
    print(f"Crash (double brackets): {e}")
except KeyError as e:
    print(f"Crash (double brackets, pandas 3.x): {e}")

try:
    df_new = df[list]
except KeyError as e:
    print(f"Crash (missing column): {e}")
